libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/Scopes/ExpressionPropertyBinder.cs

```diff
@@ -0,0 +1,142 @@
+{
+    /// <summary>
+    /// ExpressionPropertyBinder is a wrapper around any object to support expression binding.
+    /// </summary>
+    /// <remarks>
+    /// The ExpressionPropertyBinder provides read-only dictionary semantics for getting access to properties of an object.
+    /// If the value of a property is an IExpressionProperty, then the expression property will be evaluated using the DC.
+    /// Any complex objects that are returned from this are further wrapped in an ExpressionPropertyBinder, so that you can
+    /// get ExpressionProperty binding for properties in a complex hierarchy of objects.
+    /// </remarks>
+    internal class ExpressionPropertyBinder : IDictionary<string, object>
+    {
+        private const string NOTSUPPORTED = "Changing dialog definitions at run time is not supported";
+        private DialogContext dc;
+        private object obj;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExpressionPropertyBinder"/> class.
+        /// </summary>
+        /// <param name="dialogContext">dialog context for evalutation of expression properties.</param>
+        /// <param name="obj">object to wrap.  Any expression properties on it will be evaluated using the dc.</param>
+        public ExpressionPropertyBinder(DialogContext dialogContext, object obj)
+        {
+            this.dc = dialogContext;
+            this.obj = obj;
+        }
+
+        public object this[string key] { get => GetValue(key); set => throw new NotSupportedException(NOTSUPPORTED); }
+
+        public ICollection<string> Keys => ObjectPath.GetProperties(this.obj).ToList();
+
+        public ICollection<object> Values => ObjectPath.GetProperties(this.obj).Select(propertyName => GetValue(propertyName)).ToList();
+
+        public int Count => ObjectPath.GetProperties(this.obj).Count();
+
+        public bool IsReadOnly => true;
+
+        public void Add(string key, object value)
+        {
+            throw new NotSupportedException(NOTSUPPORTED);
+        }
+
+        public void Add(KeyValuePair<string, object> item)
+        {
+            throw new NotSupportedException(NOTSUPPORTED);
+        }
+
+        public void Clear()
+        {
+            throw new NotSupportedException(NOTSUPPORTED);
+        }
+
+        public override bool Equals(object other)
+        {
+            if (other is ExpressionPropertyBinder esp)
+            {
+                return this.obj.Equals(esp.obj);
+            }
+
+            return false;
+        }
+
+        public override int GetHashCode()
+        {
+            return this.obj.GetHashCode();
+        }
+
+        public bool Contains(KeyValuePair<string, object> item)
+        {
+            throw new NotSupportedException(NOTSUPPORTED);
+        }
+
+        public bool ContainsKey(string key)
+        {
+            return ObjectPath.GetProperties(this.obj).Any(p => string.Compare(p, key, ignoreCase: true) == 0);
```

Comment:
I prefer p.Equals(key, StringComparison.InvariantCultureIgnoreCase)) but this is OK too